### PR TITLE
Add support for using MS Graph instead of AAD

### PR DESCRIPTION
**Pull request by pcman312**
# Overview
The [Azure Active Directory Graph API (AAD) is being deprecated in favor of the Microsoft Graph API](https://docs.microsoft.com/en-us/graph/migrate-azure-ad-graph-faq). The AAD API is slated to be removed from Azure mid-2022. This PR updates the secrets engine to support using either API so users can migrate to using MS Graph.

**NOTE:** Since the AAD API is going to be removed entirely, this engine will stop supporting using the AAD API at some point in the future. All users of this engine should migrate to using the MS Graph API. The exact date for the removal of the AAD API from this engine has not be set.

Big thanks to @mdgreenfield for doing a large part of the heavy lifting for this PR

# Design of Change
- Adds a new config field (`use_microsoft_graph_api`) that toggles between using AAD and MS Graph
- The provider has been updated so it can switch between using the AAD API (via the `graphrbac` library) and using the `autorest` library to use the MS Graph API
- The API logic has been extracted to a separate package (`api`)

# Related Issues/Pull Requests
https://github.com/hashicorp/vault-plugin-secrets-azure/pull/63
https://github.com/hashicorp/vault-plugin-secrets-azure/pull/59

# Contributor Checklist
- [ ] Add relevant docs to upstream Vault repository, or sufficient reasoning why docs won’t be added yet
  - TBD
- [x] Backwards compatible
- [x] Add output for any tests not ran in CI to the PR description (eg, acceptance tests)
<details>
<summary>Test Output</summary>

AAD:

```
$ VAULT_ACC=1 richgo test -v -run TestCredentialInteg_aad
START| CredentialInteg_aad
START|   CredentialInteg_aad/service_principals
     | === PAUSE TestCredentialInteg_aad/service_principals
START|   CredentialInteg_aad/static_service_principals
     | === PAUSE TestCredentialInteg_aad/static_service_principals
     | === CONT  TestCredentialInteg_aad/service_principals
     | === CONT  TestCredentialInteg_aad/static_service_principals
PASS | CredentialInteg_aad (0.00s)
PASS |   CredentialInteg_aad/static_service_principals (57.65s)
PASS |   CredentialInteg_aad/service_principals (82.23s)
PASS
PASS | github.com/hashicorp/vault-plugin-secrets-azure 82.446s
```

MS Graph:

```
$ VAULT_ACC=1 richgo test -v -run TestCredentialInteg_msgraph
START| CredentialInteg_msgraph
START|   CredentialInteg_msgraph/service_principals
     | === PAUSE TestCredentialInteg_msgraph/service_principals
     | === CONT  TestCredentialInteg_msgraph/service_principals
PASS | CredentialInteg_msgraph (0.00s)
PASS |   CredentialInteg_msgraph/service_principals (30.54s)
PASS
PASS | github.com/hashicorp/vault-plugin-secrets-azure 30.753s
```
</details>

### CODE REVIEW

File: api/groups.go [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+	AddGroupMember(ctx context.Context, groupObjectID string, memberObjectID string) error
+	RemoveGroupMember(ctx context.Context, groupObjectID, memberObjectID string) error
+	GetGroup(ctx context.Context, objectID string) (result ADGroup, err error)
+	ListGroups(ctx context.Context, filter string) (result []ADGroup, err error)
+}
+
+type ADGroup struct {
+	ID          string
+	DisplayName string
+}
```
Comment (jasonodonnell) [inline api/groups.go:17]: This interface is returning `ADGroup`, which seems specific to AAD. Is this client also not being used for msgraph?

Comment (pcman312) [reply api/groups.go:17]: I don't think this is specific to AAD. I created this struct (as opposed to moving it from the main package) so I could create the interface that abstracts the differences between AAD and msgraph. If I recall correctly, the interface did exist, but it just matched the AAD client's interface but I changed it so it was abstracted away from the AAD implementation (which also meant I needed to add this `ADGroup` struct). Unless I'm misunderstanding your question?

Comment (jasonodonnell) [reply api/groups.go:17]: Can we change the name of ADGroup, then?

**Review by jasonodonnell — APPROVED** [review]
LGTM and tested well!

**Pull request merged**